product/util.py

```diff
@@ -154,15 +157,52 @@ def merge_products(left_product, right_product):
     return merged
 
 
-def diff_products(left_product, right_product, ignore_component_names=()):
+@dataclasses.dataclass
+class ComponentDiff:
+    cidentities_only_left: set = dataclasses.field(default_factory=set)
+    cidentities_only_right: set = dataclasses.field(default_factory=set)
+    cpairs_version_changed: list = dataclasses.field(default_factory=list)
+    # only set when new component is added/removed
+    names_only_left: set = dataclasses.field(default_factory=set)
+    names_only_right: set = dataclasses.field(default_factory=set)
+    # only set on update
+    names_version_changed: set = dataclasses.field(default_factory=set)
+
+
+# for v2 =  pass cd v2 and resolve components
+# make use of caching
+def diff_products(
+    left_product: cm.ComponentDescriptor,
+    right_product: cm.ComponentDescriptor,
+    ignore_component_names=(),
+    cache_dir=None,
+) -> ComponentDiff:
+    if type(left_product) is not cm.ComponentDescriptor:
+        raise NotImplementedError(
+            f'left product unsupported type {type(left_product)=}.'
+            ' Only ComponentDesciptorV2 is supported',
+        )
+    if type(right_product) is not cm.ComponentDescriptor:
+        raise NotImplementedError(
+            f'unsupported type {type(right_product)=}.'
+            ' Only ComponentDesciptorV2 is supported',
+        )
     # only take component references into account for now and assume
     # that component versions are always identical content-wise
-    left_components = {
-        c for c in left_product.components() if c.name() not in ignore_component_names
-    }
-    right_components = {
-        c for c in right_product.components() if c.name() not in ignore_component_names
-    }
+    left_components: typing.Generator[cm.Component] = product.v2.components(
+        component_descriptor_v2=left_product,
+        cache_dir=cache_dir,
+    )
+    right_components: typing.Generator[cm.Component] = product.v2.components(
+        component_descriptor_v2=right_product,
+        cache_dir=cache_dir,
+    )
+    left_components = tuple(
+        c for c in left_components if c.name not in ignore_component_names
+    )
+    right_components = tuple(
+        c for c in right_components if c.name not in ignore_component_names
+    )
 
     return diff_components(
```

Comment:
mhh..... so all `diff_products` does on top of `diff_components` is resolving the transitive closure of components? hmmmm :-/